### PR TITLE
Optimise View.asList() side inputs for iterating rather than for indexing.

### PR DESCRIPTION
**Pull request by robertwb**
The current implementation is, essentially, a distributed hashmap from integer keys to the list contents, mediated by each upstream worker starting at a random value to minimize overlaps and emitting sufficient metadata to map this onto the contiguous range [0, N). This provides optimal *random-access* performance, but very poor *iteration* performance (essentially having to do a key lookup for every advance, and as the keys are hashed and distributed rather than clustered numerically, there is little to no amortiziation in these lookups for adjacent items.

Given that most uses for List side inpupts are merely to gather a collection of values (the user has no control over the ordering when materialized) and the high costs of providing random access, this is probably the wrong tradeoff for most pipelines.

This is an update-incompatable change and so has been guarded by the update compatibility version flag. The old behavior can be explicilty asked for via a new AsList#withRandomAccess() method.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @kennknowles @damccorm

**Comment by kennknowles** [comment]
Totally agree. I do know that this was actually an explicit decision. The history as I understand it:

 - We already had `View.asIterable` that was a simple iterator, but windowed side inputs had awful performance because it was just a filter on the whole side input
 - We added `View.asList` primarily as an indicator that the per-window value could be cached in memory after the first read.
 - We added ISM format and for some reason emphasized the random access behavior of the Java `List` class.

TBH I would be perfectly happy if we had never allowed random access for list side inputs, leaving that to map side inputs.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java [hunk 1]
```diff
@@ -960,6 +995,179 @@ public void verifyDeterministic() throws NonDeterministicException {
...
   }
 
+  /**
+   * Implementation which is able to adapt an iterable materialization to a {@code List<T>}.
+   *
+   * <p>Unlike ListViewFn2, this implementation is optimized for iteration rather than indexing.
+   *
+   * <p>For internal use only.
+   */
+  public static class ListViewFn3<T> extends ViewFn<IterableView<T>, List<T>> {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:1005]: I don't _love_ that these are just named 1, 2, 3...

**Review by kennknowles — APPROVED** [review]
Does this make it identical to `View.asIterable`?

Is it possible to retain it as a window --> iterable map for efficient access by window? (I don't know if that is already implied by how this interacts with code elsewhere or what)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:1005]: Yes, this is equivalent to View.asIterable() plus implementing the List methods so nothing breaks. Good question about the Window -> Iterable map; this is handled at a lower level, but I don't know all the details there (though in that case I can see that constructing the mapping would be more worthwhile). In the interest of being conservative while capturing the most important gains I'll restrict this to the global window case.

**Comment by robertwb** [comment]
I think you're right. This was masked by 

```
/Users/robertwb/Work/beam/incubator-beam/sdks/java/extensions/avro/build/generated-test-avro-java/org/apache/beam/sdk/extensions/avro/schemas/TestAvro.java:135: error: cannot find symbol
  protected static final org.apache.avro.data.TimeConversions.TimestampConversion TIMESTAMP_CONVERSION = new org.apache.avro.data.TimeConversions.TimestampConversion();
```

**Comment by robertwb** [comment]
> This seems a likely culprit for the failure at https://ge.apache.org/s/6b22rnlopcdxk/tests/overview?outcome=FAILED

https://github.com/apache/beam/pull/31149

**Comment by Abacn** [comment]
This also breaks several Java PostCommit: https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_DataflowV1.yml?query=event%3Aschedule

faling tests:

```
testSequentialWrite (org.apache.beam.sdk.io.gcp.spanner.SpannerWriteIT) failed

java.lang.NullPointerException: Unknown producer for value SimplePCollectionView{tag=Tag<org.apache.beam.sdk.values.PCollectionViews$SimplePCollectionView.<init>:1443#e1789f47d74ca86c>, viewFn=org.apache.beam.sdk.values.PCollectionViews$IterableBackedListViewFn@b96f9be, coder=VoidCoder, windowMappingFn=GlobalWindowMappingFn{}, pCollection=wait/To wait view 0/Sample.Any/Flatten.Iterables/FlattenIterables/FlatMap/ParMultiDo(Anonymous).output [PCollection@1669047811]} while translating step wait/Wait/Map
	at org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:1269)
	at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$Translator.getProducer(DataflowPipelineTranslator.java:571)
```

```
testFhirIO_Import[DSTU2] (org.apache.beam.sdk.io.gcp.healthcare.FhirIOWriteIT) failed
testFhirIO_Import[STU3] (org.apache.beam.sdk.io.gcp.healthcare.FhirIOWriteIT) failed

java.lang.NullPointerException: Unknown producer for value SimplePCollectionView{tag=Tag<org.apache.beam.sdk.values.PCollectionViews$SimplePCollectionView.<init>:1443#53e357e8e9199108>, viewFn=org.apache.beam.sdk.values.PCollectionViews$IterableBackedListViewFn@7bc0659a, coder=VoidCoder, windowMappingFn=GlobalWindowMappingFn{}, pCollection=FhirIO.Write/FhirIO.Import/Wait On File Writing/To wait view 0/Sample.Any/Flatten.Iterables/FlattenIterables/FlatMap/ParMultiDo(Anonymous).output [PCollection@876681835]} while translating step FhirIO.Write/FhirIO.Import/Wait On File Writing/Wait/Map
	at org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:1269)
	at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$Translator.getProducer(DataflowPipelineTranslator.java:569)
	at org.apache.beam.runners.dataflow.DataflowPipelineTranslator.translateSideInputs(DataflowPipelineTranslator.java:1205)

```

**Pull request merged**